flybywiresim/aircraft
# refactor: phase 2 of rust payload, airframe and fuel module to enhance performance, optimise and support multiple-gate boarding

Payload EFB loading now has a proper frontend/backend architecture with full unit testing and better performance with less logic operations in JavaScript/TypeScript. 



Fixes #[issue_no]

## Summary of Changes

* Moved all CG MAC calculations and associated logic into rust
* Added airframe.rs module with tests
* Modified fuel.rs module with tests
* Calculate current position of CG in 3 dimensions x y and z in systems.wasm, and enabled piping into other modules
* Read all fuel levels and calculate fuel CG in rust and enabled piping into other modules
* Split payload components into more functional components
* Split backend payload into more modules and structs (i.e. for GSX logic) which can be shared
* Removed all logic for calculating CG MAC in Payload.tsx
* Removed redundant variables and lists
* Added preliminary logic for multiple passenger cabin deck support.
* Added placeholder support for a certain larger airbus
* Added backend (but not frontend in the EFB) support for multiple deck boarding - To be finished in a future PR

## Screenshots (if necessary)




## References


* A320 Aircraft Characteristics and Maintenance Planning manual (AC) 
* A380 Aircraft Characteristics and Maintenance Planning manual (AC) 




## Additional context

Future phase 3 and beyond possibilities:
* Using new JS<->WASM pipe with JSON to reduce simvar usage further and optimise (pending performance analysis)
* Extending cargo as a deck with ULDs and proper batch loading support/realistic ground ops
* Moving EFB Payload frontend to avionics framework


Discord username (if different from GitHub):2Cas#1022

## Testing instructions


1. Go to Ground > Services > Payload page, 
2. Adjust seating by clicking on the seating layout. This should add new seats and increase the table value for planned passengers.
3. Add baggage by clicking on the sliders, this should affect the desired cargo value in the table.
4. Change per passenger and per bag weight from default to higher/lower values (or leave at default)
5. Input desired passenger and desired cargo in the table
6. Check that ZFW calculation is accurate (Empty Weight is 42500KG, add passenger, passenger bags and cargo weight to this value and this should match desired ZFW). You can also verify this (approximately) with the MSFS Fuel and Weights Menu.
7. Click on the swap icon to display GW, and check that GW calculation is accurate.
9. Board passengers with desired time setting (Real/Fast/Instant). Should be able to change speed while passengers are boarding and have it transition between boarding times seamlessly.
10. Deboard passengers (either by inputting in the table, or clicking on the seating layout, then clicking the Begin Boarding button).
11. Go to the main EFB dashboard by clicking the FBW Logo in the top left, import a simbrief OFP
12. Go back to the payload page and click the cloud icon next to the table to load the OFP. Check that all relevant table values are correctly imported.
13. Load passengers/cargo again
14. Fill out INIT-A and INIT-B in MCDU (Init fuel pred). Check that MLDW indicator now shows on the performance envelope chart.
15. Proceed with normal flight operations. Check that when in air, MTOW envelope is replaced by the FLIGHT envelope.
16. Check that visuals in all themes (dark/light/blue) are correct and readable, with both bright daytime and at night (in sim time)
17. Upon landing, use the deboard button to deboard the flight with either REAL or FAST deboarding. This will open a pop-up modal which should be dismissed, and the deboarding should commence. This should work the first time to be considered a pass. 


## How to download the PR for QA

Every new commit to this PR will cause a new A32NX artifact to be created, built, and uploaded.

1. Make sure you are signed in to GitHub
1. Click on the **Checks** tab on the PR
1. On the left side, click on the bottom **PR** tab
1. Click on the **A32NX** download link at the bottom of the page
